src/charm.py

```diff
@@ -3,258 +3,112 @@
-    """Defines valid attributes of the stored state for the Jenkins Agent.
-
-    Attrs:
-       relation_configured: Whether the relation is configured.
-       jenkins_url: Jenkins server URL.
-       relation_agent_name: Agent name.
-       relation_agent_token: Token for this agent.
-    """
-
-    # Disabling since class is used to add type information to the stored state.
-    # pylint: disable=too-few-public-methods
-
-    relation_configured: bool | None
-    jenkins_url: str | None
-    relation_agent_name: str
-    relation_agent_token: str | None
+import agent
+import pebble
+import server
+from state import AGENT_RELATION, SLAVE_RELATION, InvalidStateError, State
 
+logger = logging.getLogger()
 
-class JenkinsAgentEnvConfig(typing.TypedDict):
-    """The environment configuration for the jenkins agent.
-
-    Attrs:
-        JENKINS_AGENTS: List of Jenkins agents.
-        JENKINS_TOKENS: List of Jenkins tokens.
-        JENKINS_URL: Jenkins server URL.
-    """
-
-    JENKINS_AGENTS: str
-    JENKINS_TOKENS: str
-    JENKINS_URL: str
-
-
-class JenkinsAgentCharm(CharmBase):
-    """Charm for Jenkins Agent on kubernetes.
-
-    Attrs:
-        service_name: Jenkins service name.
-    """
 
-    _stored = JenkinsAgentCharmStoredState()
-    service_name = "jenkins-agent"
+class JenkinsAgentCharm(ops.CharmBase):
+    """Charm Jenkins k8s agent."""
 
-    def __init__(self, *args) -> None:
-        """Initialize the instance.
+    def __init__(self, *args: typing.Any):
+        """Initialize the charm and register event handlers.
 
         Args:
-            args: Arguments for the CharmBase superclass.
+            args: Arguments to initialize the charm base.
         """
         super().__init__(*args)
-        self.framework.observe(self.on.start, self._on_config_changed)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.upgrade_charm, self._on_config_changed)
-        self.framework.observe(self.on.slave_relation_joined, self._on_agent_relation_joined)
-        self.framework.observe(self.on.slave_relation_changed, self._on_agent_relation_changed)
+        try:
+            self.state = State.from_charm(self)
+        except InvalidStateError as exc:
+            self.unit.status = ops.BlockedStatus(exc.msg)
+            return
 
-        self._stored.set_default(
-            relation_configured=False,
-            jenkins_url=None,
-            relation_agent_name=f"{self.unit.name.replace('/', '-')}-{uuid.uuid4()}",
-            relation_agent_token=None,
-        )
+        self.pebble_service = pebble.PebbleService(self, self.state)
+        self.agent_observer = agent.Observer(self, self.state, self.pebble_service)
 
-    def _get_pebble_config(self) -> dict:
-        """Generate the pebble config for the charm.
+        self.framework.observe(
+            self.on.jenkins_k8s_agent_pebble_ready, self._register_agent_from_config
```

Comment:
Removed the pebble ready event since it doesn't quite do anything and config changed event will trigger necessary actions. Added a handler that matches the name for each event. Thank you!

Reply:
I think I found back the issue,  when a pod is rescheduled and ends up on the same host, the upgrade-charm event is not triggered (this hapenned on the redis / indico issue).
But I think the config-changed will be triggered anyway. So we should be fine here

Reply:
I see, thank you for the update!